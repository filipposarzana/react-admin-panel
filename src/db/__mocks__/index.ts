import { User } from '~/db'

export const users: User[] = [
  {
    friends: [
      {
        name: 'A Friend',
        uuid: 'acc82963-61c3-37ec-b334-e8f6bf6b7f57',
      },
    ],
    name: 'A User',
    uuid: '6bb0295f-8665-30a1-af76-6d1236b88971',
  },
  {
    friends: [],
    name: 'A User Friend',
    uuid: 'acc82963-61c3-37ec-b334-e8f6bf6b7f57',
  },
]
