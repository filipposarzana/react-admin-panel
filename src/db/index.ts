export type BaseUser = {
  name: string
  uuid: string
}

export type User = BaseUser & {
  friends: BaseUser[]
}

export type UpdatableUser = Omit<User, 'uuid'>

export const users: User[] = []
