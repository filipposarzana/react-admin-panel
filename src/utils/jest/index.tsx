import { render } from '@testing-library/react'
import { renderHook } from '@testing-library/react-hooks'
import { ReactElement } from 'react'
import { createRenderer } from 'react-test-renderer/shallow'
import { StyleSheetManager } from 'styled-components'

export { act, fireEvent, waitFor } from '@testing-library/react'

export const renderer = {
  deep: (node: ReactElement) => {
    const r = render(<StyleSheetManager disableVendorPrefixes>{node}</StyleSheetManager>)

    return {
      ...r,
      output: r.container.firstChild,
    }
  },
  hook: <T extends unknown, R extends unknown>(hookFunction: (props: T) => R) => {
    const r = renderHook(hookFunction)

    return r
  },
  shallow: (node: ReactElement) => {
    const r = createRenderer()

    r.render(node)

    return {
      instance: r.getMountedInstance(),
      output: r.getRenderOutput(),
      unmount: r.unmount,
    }
  },
}
