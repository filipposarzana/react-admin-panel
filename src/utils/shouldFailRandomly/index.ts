export const shouldFailRandomly = () => Math.round(Math.random()) === 0
