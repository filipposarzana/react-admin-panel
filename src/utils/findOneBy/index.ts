import { User, users } from '~/db'

type Key = Extract<keyof User, 'name' | 'uuid'>

const findOneBy = <K extends Key>(prop: K) => (value: User[K]) =>
  users.find((user) => user[prop].toLowerCase() === value.toLowerCase())

export const findOneByName = findOneBy('name')

export const findOneByUuid = findOneBy('uuid')
