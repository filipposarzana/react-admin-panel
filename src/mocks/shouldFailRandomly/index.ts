import { mocked } from 'ts-jest/utils'
import { shouldFailRandomly } from '~/utils/shouldFailRandomly'

export const mockedShouldFailRandomly = mocked(shouldFailRandomly)
