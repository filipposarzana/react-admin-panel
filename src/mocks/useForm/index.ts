import { mocked } from 'ts-jest/utils'
import { FormState, useForm, UseFormReturn } from 'react-hook-form'

type Values = Record<string, unknown>

const defaultFormState = {
  isSubmitted: false,
  isSubmitting: false,
  isValid: false,
}

type GetMockedFormParam = {
  formState?: Partial<FormState<Values>>
  values?: Values
}

export const getMockedForm = ({ formState, values = {} }: GetMockedFormParam = {}) =>
  (({
    formState: {
      ...defaultFormState,
      ...formState,
    },
    handleSubmit: jest.fn((cb: (data: Values) => void) => () => cb(values)),
    setValue: jest.fn(),
  } as unknown) as UseFormReturn<Values>)

export const useMockedForm = mocked(useForm)
