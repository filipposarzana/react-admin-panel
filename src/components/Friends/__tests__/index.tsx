import { users } from '~/db'
import { getMockedForm, useMockedForm } from '~/mocks/useForm'
import { fireEvent, renderer } from '~/utils/jest'
import { Friends } from '..'

jest.mock('~/db')

afterEach(jest.clearAllMocks)

const onAdded = jest.fn()

describe('Friends', () => {
  it('calls onAdded callback with added friend', async () => {
    useMockedForm.mockReturnValue(getMockedForm({ values: { friendUuid: users[2].uuid } }))

    const { getByTestId } = renderer.deep(<Friends friends={[users[1]]} onAdded={onAdded} users={users} />)

    fireEvent.submit(getByTestId('form-friends'))

    expect(onAdded).toHaveBeenLastCalledWith([users[1], users[2]])
  })

  it('does not call onAdded callback if friend is not found', async () => {
    useMockedForm.mockReturnValue(getMockedForm({ values: { friendUuid: 'f315e4b0-a921-36bf-8c5d-088c6992975d' } }))

    const { getByTestId } = renderer.deep(<Friends friends={[]} onAdded={onAdded} users={[users[0]]} />)

    fireEvent.submit(getByTestId('form-friends'))

    expect(onAdded).toHaveBeenCalledTimes(1)
    expect(onAdded).toHaveBeenNthCalledWith(1, [])
  })

  it('matches snapshot when no users', async () => {
    const { output } = renderer.shallow(<Friends friends={[]} onAdded={onAdded} users={[]} />)

    expect(output).toMatchSnapshot()
  })
})
