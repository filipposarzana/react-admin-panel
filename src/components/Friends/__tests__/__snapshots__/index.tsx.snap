// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Friends matches snapshot when no users 1`] = `null`;
